### PR TITLE
Add HTTP request parsing

### PR DESCRIPTION
**Pull request by miselin**
This extends the existing http parser to support requests as well as responses. This allows, for example, security connectors to parse HTTP requests as part of their do_handshake() method.

httpcli continues to exist and work as it has previously, though in the new directory src/core/http (to reflect the fact the directory now contains code relevant to parsing requests, which httpcli would not generally involve itself in).

### CODE REVIEW

**Comment by yang-g** [comment]
This is ok to test.

**Comment by ctiller** [comment]
Thanks for this! It might take me a day or two to get to it (my schedule is
loaded at the moment), but I will get to it.

**Comment by ctiller** [comment]
This looks really good.

There's some merge conflicts that I'm hoping you can fix, and then let's see how the tests look... but basically LGTM.

**Comment by miselin** [comment]
Thanks, I've now fixed up the merge conflicts and also fixed a couple broken asan breakages that I introduced.

**Comment by jboeuf** [comment]
@ctiller can we merge this?

**Comment by ctiller** [comment]
@nicolasnoble and @jtattermusch are solely responsible for doing merges right now - but I've added it to their queue

File: src/core/http/format_request.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2015, Google Inc.
+ * Copyright 2016, Google Inc.
```
Comment (ctiller) [inline src/core/http/format_request.h]: 2015-2016

File: src/core/http/parser.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
+/*
+ *
+ * Copyright 2016, Google Inc.
```
Comment (ctiller) [inline src/core/http/parser.h]: 2015-2016

File: test/core/http/format_request_test.c [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2015, Google Inc.
+ * Copyright 2016, Google Inc.
```
Comment (ctiller) [inline test/core/http/format_request_test.c]: 2015-2016

**Comment by ctiller** [comment]
There are a few nits for copyrights (we're not allowed to move copyrights forward, but we can extend them).

LGTM otherwise.

**Comment by miselin** [comment]
No problems, I fixed the copyrights up. Thanks.

**Comment by jtattermusch** [comment]
this is ok to test

**Pull request merged**